### PR TITLE
test: add rancher integration

### PR DESCRIPTION
**Pull request by FrankYang0529**

### CODE REVIEW

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+            )
+            if 'status' in data and 'ready' in data['status'] and data['status']['ready']:
+                break
+            sleep(5)
+        else:
+            raise AssertionError(
+                f"MgmtCluster {self.get_harvester_cluster_name(unique_name)} can't be ready \
+                    with {wait_timeout} timed out\n"
+                f"Got error: {code}, {data}"
+            )
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:105]: need to add the check point to make sure the cluster imported and **State** is `Active` in rancher.

File: apiclient/rancher_api/managers.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                "generateName": "cc-",
+                "namespace": FLEET_DEFAULT_NAMESPACE
+            },
+            "_name": name,
+            "annotations": {
+                "provisioning.cattle.io/driver": "harvester"
+            },
+            "harvestercredentialConfig": {
+                "clusterType": "imported",
+                "clusterId": cluster_id,
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:116]: `clusterType` and `clusterId` should not be constant in here, we can create **Cloud Credential** by _External Harvester_, so `clusterType` should be determined by `clusterId` (if not provided, `clusterType` should be `external`)

File: apiclient/rancher_api/managers.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                "kubeconfigContent": kubeconfig
+            },
+            "_type": "provisioning.cattle.io/cloud-credential",
+            "name": name
+        }
+
+    def create(self, name, cluster_id, kubeconfig, *, raw=False):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:123]: `cluster_id` could be empty, for creating from external Harvester.

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        harvester_kubeconfig = api_client.generate_kubeconfig()
+        code, data = rancher_api_client.cloud_credentials.create(
+            unique_name,
+            data['status']['clusterName'],
+            harvester_kubeconfig
+        )
+
+        assert 201 == code, (
+            f"Failed to create cloud credential with error: {code}, {data}"
+        )
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:120]: need to add the check point to make sure the credential is added. (by get the credential)

File: apiclient/rancher_api/managers.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    PATH_fmt = "k8s/clusters/{cluster_id}/v1/harvester/kubeconfig"
+
+    def create_data(self, name):
+        return {
+            "clusterRoleName": "harvesterhci.io:cloudprovider",
+            "namespace": DEFAULT_NAMESPACE,
+            "serviceAccountName": name
+        }
+
+    def create(self, name, cluster_id, *, raw=True):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:141]: update the BaseManager, then the option `raw` will align other methods.

https://github.com/harvester/tests/blob/5082bc97a20b7d40446266330dc980fffeb3af1a/apiclient/harvester_api/managers.py#L32-L45

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+
+        kubeconfig = resp.text
+        print(kubeconfig)
+
+        code, data = rancher_api_client.secrets.create(
+            name=unique_name,
+            namespace=FLEET_DEFAULT_NAMESPACE,
+            data={
+                "credential": base64.b64encode(kubeconfig.encode('UTF-8')).decode('UTF-8')
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:173]: encode should be handled in API, not user side.

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+        assert 201 == code, (
+            f"Failed to create secret with error: {code}, {data}"
+        )
+
+        code, data = rancher_api_client.harvester_configs.create(
+            name=unique_name,
+            cpu_count="2",
+            disk_size="40",
+            memory_size="4",
+            image_name=f"{DEFAULT_NAMESPACE}/{unique_name}",
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:190]: namespace should be determined by the image, not here.

here is the example: https://github.com/harvester/tests/blob/043f06eaa38e6dd114ceb10a1bbfedbd02ec2dc8/harvester_e2e_tests/integration/test_vm_functions.py#L34-L56

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+                cloud_credential = d
+                break
+
+        assert cloud_credential is not None, (
+            f"Failed to find cloud credential: {data}"
+        )
+
+        code, data = rancher_api_client.mgmt_clusters.create(
+            self.get_rke2_cluster_name(unique_name),
+            spec={
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:218]: spec should be placed in API client.

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+                    "enabled": False,
+                    "caCerts": "",
+                    "fqdn": ""
+                }
+            }
+        )
+
+        assert 201 == code, (
+            f"Failed to create RKE2 MgmtCluster {unique_name} with error: {code}, {data}"
+        )
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:313]: need to add check points to make sure VMs are created, provisioning, and states.

**Review by lanfon72 — CHANGES_REQUESTED** [review]

File: harvester_e2e_tests/apis/test_settings.py [hunk 10]
```diff
@@ -60,3 +60,16 @@ def test_get_all_settings_v110(api_client, expected_settings):
...
+@pytest.mark.p0
+@pytest.mark.settings
+def test_update_log_level(api_client):
+    updates = {"value": "Debug"}
+    code, data = api_client.settings.update("log-level", updates)
+
+    assert 200 == code, (f"Failed to update log-level setting with error: {code}, {data}")
+
+    # For teardown
+    updates = {"value": "Info"}
```
Comment (lanfon72) [inline harvester_e2e_tests/apis/test_settings.py:74]: When doing update test, steps would like:
1. Get original values
2. Update to expected values
3. Check values been updated
4. Update to original values

File: apiclient/rancher_api/managers.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    CREATE_fmt = "v1/provisioning.cattle.io.clusters"
+    GET_fmt = "apis/provisioning.cattle.io/v1/namespaces/{ns}/clusters/{uid}"
+    DELETE_fmt = "v1/provisioning.cattle.io.clusters/{ns}/{uid}"
+
+    def create_data(self, cluster_name, spec=None, labels=None):
+        if spec is None:
+            spec = {}
+
+        if labels is None:
+            labels = {}
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:81]: one of the simply way to do the check is:
```python
labels = labels or {}
# when `dict(1=2) or {}`, return former
# when `None or {}`, return latter
# when `"" or {}`, return latter
```
by using the tip, we can just put them in the return data, like:
```python
# ...
"labels": labels or {}
```

File: apiclient/rancher_api/managers.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+            },
+            "networkName": network_id,
+            "sshUser": ssh_user,
+            "userData": base64.b64encode(user_data.encode('UTF-8')).decode('UTF-8'),
+            "vmNamespace": vm_namespace,
+            "type": "rke-machine-config.cattle.io.harvesterconfig"
+        }
+
+    def create(self, name, cpus, mems, disks, image_id, network_id,
+               ssh_user, vm_namespace=DEFAULT_NAMESPACE, user_data="", *, raw=False):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:205]: I just awared that we forgot there still have `network_data`? :thinking:

File: harvester_e2e_tests/apis/test_settings.py [hunk 13]
```diff
@@ -60,3 +60,20 @@ def test_get_all_settings_v110(api_client, expected_settings):
...
         warnings.warn(UserWarning(f"New setting(s) added: {added}"))
+
+
+@pytest.mark.p0
+@pytest.mark.settings
+def test_update_log_level(api_client):
+    code, data = api_client.settings.get("log-level")
+    assert 200 == code, (f"Failed to get log-level setting with error: {code}, {data}")
+
+    original_value = data.get("value", "")
```
Comment (lanfon72) [inline harvester_e2e_tests/apis/test_settings.py:71]: The value should not have any chance to be empty :thinking:

File: apiclient/rancher_api/managers.py [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+            raw=raw
+        )
+
+    def create(self, name, cloud_provider_config_id, hostname_prefix,
+               harvester_config_name, k8s_version, cloud_credential_id, raw=False):
+        data = self.create_data(name, cloud_provider_config_id, hostname_prefix,
+                                harvester_config_name, k8s_version, cloud_credential_id)
+        return self._create(self.PATH_fmt.format(uid="", ns=""), json=data, raw=raw)
+
+    def create_harvester(self, name, raw=False):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:190]: asterisk(`*`) is necessary, it means argument `raw` is keyword-only argument.

File: apiclient/rancher_api/managers.py [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+            "harvestercredentialConfig": harvester_credential_config,
+            "_type": "provisioning.cattle.io/cloud-credential",
+            "name": name
+        }
+
+    def create(self, name, kubeconfig, cluster_id="", *, raw=False):
+        data = self.create_data(name, kubeconfig, cluster_id)
+        return self._create(self.PATH_fmt.format(uid="", ns=""), json=data, raw=raw)
+
+    def get(self, name="", raw=False):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:253]: asterisk

File: apiclient/rancher_api/managers.py [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+        if name == "":
+            return self._get(self.PATH_fmt.format(uid=""), raw=raw)
+        return self._get(self.PATH_fmt.format(uid=f"/{name}"), raw=raw)
+
+    def delete(self, name):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:258]: method signature should add `raw=False`

File: apiclient/rancher_api/managers.py [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+                "sshPassword": "",
+                "sshPort": "22",
+                "sshPrivateKeyPath": "",
+                "sshUser": ssh_user,
+                "userData": user_data,
+                "vmAffinity": "",
+                "vmNamespace": vm_namespace,
+                "type": "harvesterConfig"
+            },
+            "namespaceId": "fixme",
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:377]: is the `fixme` need to be fix or it's real parameter to be use? :thinking:

File: apiclient/rancher_api/managers.py [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+            "cloudCredentialId": cloud_credential_id,
+            "labels": {},
+            "name": name
+        }
+
+    def get(self, name="", *, raw=False):
+        if name == "":
+            return self._get(self.PATH_fmt.format(uid=""), raw=raw)
+        return self._get(self.PATH_fmt.format(uid=name), raw=raw)
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:386]: looks `if` condition is not necessary.

File: apiclient/rancher_api/managers.py [hunk 19]
```diff
@@ -0,0 +1,554 @@
...
+                "enabled": False,
+                "scheduleConfig": None,
+                "scanConfig": None
+            }
+        }
+
+    def get(self, name="", *, raw=False):
+        return self._get(self.PATH_fmt.format(uid=name), raw=raw)
+
+    def create(self, name, k8s_version, raw=False):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:518]: asterisk

File: apiclient/rancher_api/managers.py [hunk 20]
```diff
@@ -0,0 +1,554 @@
...
+            "nodeTemplateId": node_template_id,
+            "hostnamePrefix": hostname_prefix
+        }
+
+    def get(self, name="", ns="", *, raw=False):
+        if name == "":
+            return self._get(self.PATH_fmt.format(uid="", ns=""), raw=raw)
+        return self._get(self.PATH_fmt.format(uid=f":{name}", ns=ns), raw=raw)
+
+    def create(self, cluster_id, node_template_id, hostname_prefix, quantity=1, raw=False):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:549]: asterisk

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+    api_client.networks.delete(network_name)
+
+
+@pytest.fixture(scope="session")
+def focal_image_url():
+    return "http://cloud-images.ubuntu.com/releases/focal/release/ubuntu-20.04-server-cloudimg-amd64-disk-kvm.img"  # noqa
+
+
+@pytest.fixture(scope="class")
+def focal_image(request, api_client, unique_name, focal_image_url, wait_timeout):
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:74]: `request` is not used

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 22]
```diff
@@ -0,0 +1,355 @@
...
+    else:
+        raise AssertionError(
+            f"MgmtCluster {harvester_cluster_name} can't be ready \
+                with {wait_timeout} timed out\n"
+            f"Got error: {code}, {data}"
+        )
+
+    yield data
+
+    rancher_api_client.mgmt_clusters.delete(harvester_cluster_name)
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:163]: we should also remove `cluster-registration-url` in harvester side.

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+        cloud_provider_config_id = f"{data['metadata']['namespace']}:{data['metadata']['name']}"
+
+        code, data = rancher_api_client.harvester_configs.create(
+            name=unique_name,
+            cpus="2",
+            mems="4",
+            disks="40",
+            image_id=focal_image,
+            network_id=vlan_network['id'],
+            ssh_user="ubuntu",
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:234]: ssh user should not fix in here.
I would suggest to update the focal_image fixture to:
```python
yield dict(ssh_user="ubuntu", id=f"{namespace}/{name}")
```

then if we update the image, the **ssh_user** will be update accordingly.

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 24]
```diff
@@ -0,0 +1,355 @@
...
+            user_data="""
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:235]: I will suggest to use the format like:
```python
    user_data = (
        "#cloud-config\n"
        "password: test\n"
        "chpasswd:\n"
        "    expire: false\n"
        "ssh_pwauth: true\n"
    )
````

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 25]
```diff
@@ -0,0 +1,355 @@
...
+            hostname_prefix=f"{unique_name}-",
+            harvester_config_name=unique_name,
+            k8s_version=k8s_version,
+            cloud_credential_id=harvester_cloud_credential['id']
+        )
+        assert 201 == code, (
+            f"Failed to create RKE2 MgmtCluster {unique_name} with error: {code}, {data}"
+        )
+
+        endtime = datetime.now() + timedelta(seconds=wait_timeout*10)
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:259]: we should use rancher's wait_timeout

File: apiclient/rancher_api/managers.py [hunk 26]
```diff
@@ -0,0 +1,554 @@
...
+                    },
+                    "machinePools": [
+                        {
+                            "name": "pool1",
+                            "etcdRole": True,
+                            "controlPlaneRole": True,
+                            "workerRole": True,
+                            "hostnamePrefix": hostname_prefix,
+                            "labels": {},
+                            "quantity": 1,
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:151]: machines' quantity should be the argument

**Review by lanfon72 — CHANGES_REQUESTED** [review]

File: apiclient/rancher_api/managers.py [hunk 27]
```diff
@@ -0,0 +1,564 @@
...
+            "labels": {},
+            "name": name
+        }
+
+    def get(self, name="", *, raw=False):
+        return self._get(self.PATH_fmt.format(uid=name), raw=raw)
+
+    def create(self, name, cpus, mems, disks, image_id, network_id,
+               ssh_user, cloud_credential_id, vm_namespace=DEFAULT_NAMESPACE,
+               user_data="", network_data="", raw=False):
```
Comment (lanfon72) [inline apiclient/rancher_api/managers.py:390]: asterisk

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 28]
```diff
@@ -0,0 +1,365 @@
...
+        endtime = datetime.now() + timedelta(seconds=rancher_wait_timeout)
+        while endtime > datetime.now():
+            code, data = rancher_api_client.mgmt_clusters.get(rke2_cluster_name)
+            if data.get('status', {}).get('ready', False):
+                break
+            sleep(5)
+        else:
+            raise AssertionError(
+                f"RKE2 MgmtCluster {rke2_cluster_name} can't be ready \
+                    with {wait_timeout} timed out\n"
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:277]: there should update to `rancher_wait_timeout`

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 29]
```diff
@@ -0,0 +1,365 @@
...
+        endtime = datetime.now() + timedelta(seconds=rancher_wait_timeout)
+        while endtime > datetime.now():
+            code, data = rancher_api_client.mgmt_clusters.get(cluster_id)
+            if code == 200 and data.get('status', {}).get('ready', False):
+                break
+            sleep(5)
+        else:
+            raise AssertionError(
+                f"RKE1 MgmtCluster {cluster_id} can't be ready \
+                    with {wait_timeout} timed out\n"
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:347]: there should update to `rancher_wait_timeout`

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 30]
```diff
@@ -0,0 +1,365 @@
...
+                f"Got error: {code}, {data}"
+            )
+
+        # teardown
+        rancher_api_client.mgmt_clusters.delete(cluster_id)
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:353]: I'm thinking maybe we can just use `rancher_wait_timeout` for all waiting cases :thinking:

**Review by lanfon72 — CHANGES_REQUESTED** [review]

**Comment by lanfon72** [comment]
@khushboo-rancher, for RKE1 and RKE2 existing manual test cases:
- [20-Create RKE1 Kubernetes Cluster](https://harvester.github.io/tests/manual/harvester-rancher/20-create-rke1-kubernetes-cluster/)
- [21-Delete RKE1 Kubernetes Cluster](https://harvester.github.io/tests/manual/harvester-rancher/21-delete-rke1-kubernetes-cluster/)
- [22-Create RKE2 Kubernetes Cluster (e2e\_be)](https://harvester.github.io/tests/manual/harvester-rancher/22-create-rke2-kubernetes-cluster/)
- [23-Delete RKE2 Kubernetes Cluster (e2e\_be)](https://harvester.github.io/tests/manual/harvester-rancher/23-delete-rke2-kubernetes-cluster/)

Both are testing with `1` VM (as the PR covered), but I'm not sure should add more test cases to test with `3` VMs?

**Comment by khushboo-rancher** [comment]
It's good to have 3 VMs test case as well, we can add with priority level 2

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 31]
```diff
@@ -0,0 +1,409 @@
...
+    return f"{unique_name}-rke2"
+
+
+@pytest.fixture(scope='session')
+def vlan_network(request, api_client):
+    vlan_nic = request.config.getoption('--vlan-nic')
+    vlan_id = request.config.getoption('--vlan-id')
+    # don't create network if VLAN is not correctly specified
+    if vlan_id == -1:
+        return
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:56]: I'm thinking there might have a hidden issue if the vlan will not be created and still pass the fixture setup 🤔

**Review by lanfon72 — APPROVED** [review]
LGTM, @FrankYang0529 we might need run a test on chiefs to see whether this will work on our daily test environment.

**Comment by FrankYang0529** [comment]
Hi @khushboo-rancher, I already run test for `test_create_rke2` and `test_delete_rke2` in our `harvester-runtests`. Job ID is `125`. Since we need to modify `harvester-baremetal-ansible` to test RKE1, I think we can merge this PR first. We can find a convenient way to test both RKE1 and RKE2 in the future. WDYT? Thanks.

File: harvester_e2e_tests/conftest.py [hunk 32]
```diff
@@ -240,6 +240,7 @@ def pytest_configure(config):
         ('rancher_integration_with_external_rancher', (
             'mark Rancher integration tests with an external Rancher')),
         ('rancher', ('mark rancher integration  tests ')),
+        ('rke2', ('marker to run rke2 integration tests ')),
```
Comment (khushboo-rancher) [inline harvester_e2e_tests/conftest.py:243]: Are we missing rke1 marking?

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]
LGTM, just a minor marker missing.

File: harvester_e2e_tests/conftest.py (hunk 32 above)
Comment (FrankYang0529) [reply harvester_e2e_tests/conftest.py:248]: Added it. Thank you.

File: harvester_e2e_tests/integration/test_rancher_integration.py (hunk 31 above)
Comment (FrankYang0529) [reply harvester_e2e_tests/integration/test_rancher_integration.py]: Updated it. Thank you.

**Review by lanfon72 — APPROVED** [review]
LGTM

**Pull request merged**